Add a ledger event for successful Plutus scripts

JaredCorduan opened the pull request:
This PR adds a ledger event for successful Plutus scripts, providing all the information needed to re-run the scripts.

# Redux Notes for reviewers

There are two new events:
* `SuccessfulPlutusScriptsEvent [PlutusDebug]`
* `FailedPlutusScriptsEvent [PlutusDebug]`

In the case of no failures, with the transaction marked as `IsValid == True`, all the information needed to re-run all the scripts in a given transaction are a part of a `SuccessfulPlutusScriptsEvent` event. In the case of failures, with the transaction marked as `IsValid == False`, there are two events, one for the successes and one for the failures. No event is transmitted otherwise, since this triggers a predicate failure.

## Other notable changes

The type returned by `runPLCScript` (the ledger function which wraps the Plutus evaluator) is named `ScriptResult` and has been changed from

```
data ScriptResult = Passes | Fails ![FailureDescription]
```
to
```
data ScriptResult  = Passes [PlutusDebug]  | Fails [PlutusDebug] (NonEmpty ScriptFailure)
```

In other words, `ScriptResult` now collects the successful script information as well.  There is only a small difference between `FailureDescription` and `ScriptFailure`. The former contains an encoded bytestring with the needed information to re-run scripts, the later contains the un-encoded data. We cannot change the serialization of the predicates failures without breaking the node-to-client protocol, and since `FailureDescription` is used for a predicate failure we keep this type around (though it is moved to `Cardano.Ledger.Alonzo.Rules.Utxos`). The `ScriptFailure` type allows use to transmit the ledger event without encoding the data.

Lastly, the type `PlutusDebugInfo` has one small change in the `DebugInfo` constructor. Instead of the less helpful `String`, it now contains the actual Plutus error. It also now contains the `PlutusDebug`.

closes #2659

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -444,28 +444,28 @@ instance FromCBOR FailureDescription where
   fromCBOR = decode (Summands "ScriptResult" dec)
     where
-      dec 0 = SumD Passes
+      dec 0 = SumD Passes <! From
```
Note that the `Passes` constructor (unlike `Fails` below) is not a part of any predicate failure. Therefore this change to the serialization does _not_ break the node-to-client protocol.

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -444,28 +444,28 @@ instance FromCBOR FailureDescription where
-andResult ans Passes = ans
-andResult (Fails xs) (Fails ys) = Fails (xs ++ ys)
-
+-- | The ScriptResult operation is biased towards failure.
```
I guess the upshot of  this is that if I have a transaction which has some passing and some failing scripts, then I won't get any events for the succeeding scripts, only the failing ones? That's okay (and it's a weird case), just checking.

michaelpj replied:
Worth documenting perhaps.

JaredCorduan replied:
Thanks for bringing this up, it's a bit tricky and I don't know how best to handle it.

If a transaction contains one or more failing scripts, then:
* if the transaction is marked as `IsValid = True`, then the transaction will be rejected. The predicate failure will collect all the failures, and the node logs will display all the needed information to re-run the failed scripts. but indeed there is no mention of the succeeding scripts in this case.
* if the transaction is marked as `IsValid = False`, then as of right now neither the logs nor the events record this information. because of this semigroup instance, it would be easy to make an event for the failures, but an event for all the scripts would need to either re-run everything or we could redo how `evalScripts` works altogether (ie not use `<>`). I'll point out where in this PR we could put the event for the failures that are marked as failures.

michaelpj replied:
As a complete outsider, what I would have sort of expected would be that the events would be emitted as a side effect of doing the validation, and orthogonally to how the actual results of validation are reported. But I don't know if that makes any sense here :shrug:

JaredCorduan replied:
it's not quite that simple (at least not without a big refactor), but I can probably make it look more like that from the outside.

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -444,28 +444,28 @@ instance FromCBOR FailureDescription where
+  (Fails xs) <> (Fails ys) = Fails (xs <> ys)
 
 data PlutusDebug = PlutusDebug
   { debugCostModel :: CostModel,
```
Should this be a `CostModelParams`? Not sure.

JaredCorduan replied:
`CostModel` is what we currently pass to the evaluator, so that's why it is captured here.

michaelpj replied:
Is that because both `CostModel` and `CostModelParams` are synonyms for `Map Text Integer` and so they're equivalent? :thinking:

JaredCorduan replied:
Sorry, I was confused. Yes, it is essentially a `CostModelParams`. But wrapped in a newtype, and with a custom serialization.

michaelpj replied:
I guess I was expecting this to be more-or-less the arguments to the evaluation functions in the Plutus ledger api module.

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -444,28 +444,28 @@ instance FromCBOR FailureDescription where
+  fs@(Fails _) <> (Passes _) = fs
+  (Fails xs) <> (Fails ys) = Fails (xs <> ys)
 
 data PlutusDebug = PlutusDebug
```
Out of the diff, but maybe `debugExUnits` should be `debugEvaluationBudget` so it says what it is rather than just the type.

JaredCorduan replied:
sounds great!

JaredCorduan replied:
it's no longer a record, but a sum type, so bye bye accessor names :)

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -474,7 +474,7 @@ data PlutusDebug = PlutusDebug
     debugData :: [PV1.Data],
```
`debugScriptArguments`?

JaredCorduan replied:
I like it!

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -474,7 +474,7 @@ data PlutusDebug = PlutusDebug
     debugData :: [PV1.Data],
     debugVersion :: Language
```
Is this actually sufficient? If things change substantially between versions, maybe we'll need to change the structure for different languages. So we might need `PlutusV1Debug`, `PlutusV2Debug` etc. But I guess we can do that later.

michaelpj replied:
Speaking of other languages: if we're doing this, should we be consistent and emit events for success/failures of native scripts (or whatever we're calling them now)?

JaredCorduan replied:
yea, are probably right to think now about future proofing this. So I guess you are proposing using multiple constructors? that's reasonable. I'm not sure if we should do this now or when it is needed...

michaelpj replied:
I guess? In the spirit of "Plutus V1 and Plutus V2 are separate languages as far as the ledger is concerned", it seems reasonable to have a `ScriptEvaluationEvent` constructor per language. Not confident though.

JaredCorduan replied:
it's now a sum type, should be more future proof now.

michaelpj commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:
```diff
@@ -542,7 +542,7 @@ runPLCScript proxy lang (CostModel cost) scriptbytestring units ds =
     scriptbytestring
     ds of
     (_, Left e) -> explainPlutusFailure proxy lang scriptbytestring e ds (CostModel cost) units
```
How is this actually going to look in the events? From the look of this function, here we're going to get a predicate failure (?) with the _serialized_ version of `PlutusDebug`, whereas in the success case we'll just get the structure itself? It seems like that might be a bit of a pain to analyze consistently.

JaredCorduan replied:
This is all related to https://github.com/input-output-hk/cardano-ledger/pull/2670/files#r816904931

`evalScripts` is used both in the case that `IsValid` is true and false. Currently, in this PR, I've avoided the `IsValid = False` case (which is maybe a mistake) with respect to the events. This left here is currently only feeding into a predicate failure, not the events.

JaredCorduan commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -221,7 +222,7 @@ scriptsNotValidateTransition = do
       case evalScripts @era tx sLst of
-        Passes -> False ?!## ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
+        (Passes _) -> False ?!## ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
         Fails _sss -> pure ()
```
This is where we could place an event for those transactions that are successfully submitted to the chain which have a failing script and are marked as `IsValid = False`. `_sss` will collect only the failures, given the `<>` in this PR. we could redo how it all works to collect successes and failures.

michaelpj replied:
I do wonder whether this is a symptom of what I mentioned above, which is that it looks like the events and the result of checking are communicated in the same band...

JaredCorduan replied:
indeed it is. but treating the events as side effects was a more dramatic change than I had wanted to make.

Commit:
New ledger event for all plutus scripts

The event includes enough information to be able to re-run the scripts
exactly as they were executed on-chain.

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -221,8 +230,10 @@ scriptsNotValidateTransition = do
+          Passes _ps -> False ?!## ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
+          Fails ps fs ->
+            tellEvent (SuccessfulPlutusScriptsEvent ps)
+              >> tellEvent (FailedPlutusScriptsEvent (scriptFailuresToPlutusDebug fs))
```
Minor syntactic suggestion
```suggestion
          Fails ps fs -> do
            tellEvent (SuccessfulPlutusScriptsEvent ps)
            tellEvent (FailedPlutusScriptsEvent (scriptFailuresToPlutusDebug fs))
```

lehins commented on eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxos.hs:
```diff
@@ -199,8 +201,10 @@ scriptsNo = do
+        Passes _ -> False ?!## ValidationTagMismatch (getField @"isValid" tx) PassedUnexpectedly
+        Fails ps fs ->
+          tellEvent (SuccessfulPlutusScriptsEvent ps)
+            >> tellEvent (FailedPlutusScriptsEvent (scriptFailuresToPlutusDebug fs))
```
Same here:
```suggestion
        Fails ps fs -> do
          tellEvent (SuccessfulPlutusScriptsEvent ps)
          tellEvent (FailedPlutusScriptsEvent (scriptFailuresToPlutusDebug fs))
```

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -250,9 +261,36 @@ invalidEnd = intercalate "," ["[LEDGER][SCRIPTS_NOT_VALIDATE_TRANSITION]", "END"
+scriptFailureToFailureDescription :: ScriptFailure -> FailureDescription
+scriptFailureToFailureDescription (OnePhaseSF t) = OnePhaseFailure t
+scriptFailureToFailureDescription (PlutusSF t pd) =
+  PlutusFailure t (B64.encode . serialize' $ pd)
```
```suggestion
  PlutusFailure t (B64.encode $ serialize' pd)
```

JaredCorduan replied:
I keep doing that, no clue why :)

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxos.hs:
```diff
@@ -122,8 +129,10 @@ instance
+data UtxosEvent era
+  = AlonzoPpupToUtxosEvent (Event (Core.EraRule "PPUP" era))
+  | SuccessfulPlutusScriptsEvent [PlutusDebug]
+  | FailedPlutusScriptsEvent [PlutusDebug]
```
I'd suggest using `NonEmpty` here as well
```suggestion
  | FailedPlutusScriptsEvent (NonEmpty PlutusDebug)
```

This will be useful for downstream consumers of this event.

JaredCorduan replied:
I actually tried to use a `NonEmpty` list here, but ran into an unfortunate lack of "parse, don't validate". I'd love to fix it, I'm not sure how easy it will be. Maybe a follow up PR?

The function `collectTwoPhaseScriptInputs` only returns Plutus scripts (together will the collected data that it needs to run them) in the happy path, but unfortunately the returns is `AlonzoScript.Script`, thus forgetting the fact that only Plutus scripts are present.  When `evalScripts` get the results from `collectTwoPhaseScriptInputs`, it has to do something with the case of Timelocks, so it creates the possibility of a 1-phase failure, namely the constructor `OnePhaseSF Text` of `ScriptFailure`.

This leads to having to filter the non-empty list of `ScriptFailure`s to just those relating to plutus. See the function `scriptFailuresToPlutusDebug` introduced in this PR. And this is why I could not easily use a non-empty list inside `FailedPlutusScriptsEvent`.

This is a silly situation, of course, and I do want to fix it.

JaredCorduan replied:
it's worth noting that timelocks (aka "1-phase scripts") are validated in Alonzo in:

https://github.com/input-output-hk/cardano-ledger/blob/70dbfc9f12283666e29f8492aff5da9b1c8c5a7f/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs#L361

This check uses the `validateScript` class method of `ValidateScript`. One of the first bits of tech debt that I would like to address post Babbage is an overhaul in the handling of 1-phase vs 2-phase scripts. The `validateScript` methods look generic, as though it runs all scripts, but does not.

lehins replied:
It is silly indeed. I don't think it make sense to have a possibility of emitting a `FailedPlutusScriptsEvent` without any "Plutus" failure. Correct me if I am wrong, but I think there is an easy solution to this, would this work?
https://github.com/input-output-hk/cardano-ledger/pull/2724

JaredCorduan replied:
Yep, that's great. I'll use it if I can't quickly come up with a decent solution to the root of the problem.

JaredCorduan replied:
I've just pushed another commit which addresses the root of the problem, namely the unneeded checking of timelocks inside collectTwoPhaseInputs.

lehins replied:
Nice!!!! :100: 
I think this is one of those examples where a stricter type encourages a better implementation

JaredCorduan replied:
yep, absolutely. I've gotten so much mileage out of [Parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

lehins reviewed (approved):
Looks great. One suggestions for `FailedPlutusScriptsEvent` I think is worth considering. Others feel free to ignore, since they are just a matter of taste.

Commit:
Apply suggestions from code review

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

Commit:
clean up the collectTwoPhaseScriptInputs function

In particular, there is no reason to validate timelock scripts inside
collectTwoPhaseScriptInputs since this function already filters out such
scripts. This results in the ScriptFailure type being simpler.
